src/Storages/Kafka/StorageKafka2.cpp

```diff
@@ -745,63 +746,204 @@ void StorageKafka2::dropReplica()
+StorageKafka2::lockTopicPartitions(zkutil::ZooKeeper & keeper_to_use, const TopicPartitions & topic_partitions)
+{
+    const auto already_locked_partitions = lookupReplicaState(keeper_to_use);
+    for (const auto & already_locks : already_locked_partitions) {
+        LOG_INFO(
+            log,
+            "Already locked topic partition {}:{}",
+            already_locks.topic,
+            already_locks.partition_id
+        );
+    }
 
-    if (const auto code = keeper_to_use.tryMulti(ops, responses); code != Coordination::Error::ZOK)
+    for (const auto & partition_to_unlock : ReplicaState.tmp_topics_assigned)
     {
-        if (code != Coordination::Error::ZNODEEXISTS)
-            zkutil::KeeperMultiException::check(code, ops, responses);
+        unlockTopicPartition(keeper_to_use, partition_to_unlock);
+    }
 
-        LOG_TRACE(log, "Couldn't create topic partitions locks because some of them already exists");
-        // Possible optimization: check the content of lock files, if we locked them, then we can clean them up and retry to lock them.
-        return std::nullopt;
+    size_t can_lock_partitions = static_cast<size_t>(std::ceil(static_cast<double>(topic_partitions.size()) / ActiveReplicaCount));
+    TopicPartitionLocks permanent_locks;
+    TopicPartitionLocks tmp_locks;
+
+    TopicPartitions available_topic_partitions;
+    available_topic_partitions.reserve(topic_partitions.size());
+    for (const auto & partition : topic_partitions)
+    {
+        if (!already_locked_partitions.contains(partition))
+            available_topic_partitions.push_back(partition);
     }
 
-    // We have the locks, let's gather the information we needed
-    TopicPartitionLocks locks;
+    for (const auto & available_locks : available_topic_partitions) {
+        LOG_INFO(
+            log,
+            "Topic partition {}:{} is available",
+            available_locks.topic,
+            available_locks.partition_id
+        );
+    }
+
+    auto available_topics_it = available_topic_partitions.begin();
+    if (can_lock_partitions < ReplicaState.permanent_topics_assigned.size())
     {
-        auto tp_it = topic_partitions.begin();
-        auto path_it = topic_partition_paths.begin();
-        for (; tp_it != topic_partitions.end(); ++tp_it, ++path_it)
-        {
-            using zkutil::EphemeralNodeHolder;
-            LockedTopicPartitionInfo lock_info{
-                EphemeralNodeHolder::existing(*path_it / lock_file_name, keeper_to_use),
-                getNumber(keeper_to_use, *path_it / commit_file_name),
-                getNumber(keeper_to_use, *path_it / intent_file_name)};
+        TopicPartitions partitions_to_lock(ReplicaState.permanent_topics_assigned.begin(), ReplicaState.permanent_topics_assigned.begin() + can_lock_partitions);
+        for (const auto & partition_to_lock : partitions_to_lock)
+            createLocksInfo(keeper_to_use, permanent_locks, partition_to_lock);
```

Comment:
As I mentioned above, this is wrong, because you don't have "lock" the already locked partitions again.